lib/rucio/core/authentication.py

```diff
@@ -349,17 +446,856 @@ def get_auth_token_saml(account, saml_nameid, appid, ip=None, session=None):
+        account = result.account
+        nonce = result.nonce
+        access_msg = result.access_msg
+        refresh_lifetime = result.refresh_lifetime
+        auth_args = parse_qs(authn_url.query)
+        auth_args["grant_types"] = ["authorization_code"]
+
+        # re-initializing authentication request info in OIDC client (to make it work on multiple server instances)
+        OIDC_client.authorization_request_info(request_args=auth_args)
+        # parsing the authorization query string by the Rucio OIDC Client
+        OIDC_client.parse_response(AuthorizationResponse,
+                                   info=auth_query_string,
+                                   sformat="urlencoded")
+        # assembling parameters to request an access token
+        args = {"code": code,
+                "audience": auth_args['audience'],
+                "scope": auth_args['scope'],
+                }
+        record_counter(counters='IdP_authentication.code_granted')
+        # exchange access code for a access token
+        oidc_tokens = OIDC_client.do_access_token_request(state=state,
+                                                          request_args=args,
+                                                          authn_method="client_secret_basic")
+        # mitigate replay attacks
+        ID_token_nonce = oidc_tokens['id_token']['nonce']
+        if ID_token_nonce != nonce:
+            raise CannotAuthenticate("ID token could not be associated with the Rucio OIDC Client session."
+                                     + " This points to possible replay attack !")  # NOQA: W503
+
+        # recognize Rucio account using the OIDC identity string
+        identity_string = oidc_identity_string(oidc_tokens['id_token']['sub'], oidc_tokens['id_token']['iss'])
+        if account == InternalAccount('webui'):
+            try:
+                account = get_default_account(identity_string, IdentityType.OIDC, True, session=session)
+            except:
+                return ('webhome', None, 'token', None)
+        # check if given account has the identity registered
+        if session.query(models.IdentityAccountAssociation).filter_by(identity=identity_string,
+                                                                      identity_type=IdentityType.OIDC,
+                                                                      account=account).first() is None:
+            raise CannotAuthenticate("OIDC identity '%s' of the '%s' account is unknown to Rucio." % (identity_string, account))
+
+        record_counter(counters='IdP_authentication.success')
+        # get access token expiry timestamp
+        expired_at = datetime.datetime.utcnow() + datetime.timedelta(seconds=oidc_tokens['expires_in'])
+        # get audience and group information (assumes profile scope is provided) from id_token
+        audience = " ".join(auth_args['audience'])
+        at_claims = json.loads(JWT().unpack(oidc_tokens['access_token']).part[1])
```

Comment:
what if response is invalid/malformed string?

Reply:
Wrapped it in try except.